### PR TITLE
Improve setup of Databricks environment in dagster-databricks

### PR DESCRIPTION
**Pull request by sd2k**
It looks like I missed a couple of things when creating dagster-databricks.

The first is a simple bug: the `staging-prefix` field in the external step launcher needs to be an absolute path. The description even said this, but the default value wasn't absolute :see_no_evil: 

The second is something a bit more interesting! When using `StringSource` to specify config values, e.g. in the run config included below, the environment variables are available to the _initial_ dagster workers but aren't serialized and sent to the remote job which runs on Databricks. This causes the remote job to fail because those environment variables aren't available.

I can see two options:

1. Reify the config into the actual values of those environment variables when the run is first executed and before it's serialized, so that the remote PySpark executor doesn't need to look in its environment. This has the disadvantage that secrets will probably end up in the serialized step run ref.
2. Require the user to _also_ add those environment variables to Databricks secrets (as we already do for the storage configuration), and add some additional configuration which the `databricks_step_main.py` script uses to export those Databricks secrets as environment variables before the step is executed. This is the technique I've implemented in this PR. The disadvantage is that it requires duplicating the env names (once in the `StringSource` location, once in the `secrets` location), it requires the user to upload environment variables to Databricks (which, in fairness, they would have to do anyway if they wanted to submit jobs to run on Databricks), and that it doesn't help solve the issue in EMR (if it does?).

Let me know your thoughts, happy to discuss and modify this PR!

Example config which works fine when `StringSource` variables are specified concretely, not as env:
```
resources:
  adls2:
    config:
      credential:
        key:
          env: DATA_LAKE_STORAGE_ACCOUNT_KEY
      storage_account: foo
  delta_lake:
    config:
      container_name: delta
      storage_account_name: foo
  input_stream:
    config:
      connection_string:
        env: EVENTHUB_CONNECTION_STRING
      consumer_group: $Default
  pyspark:
    config:
      spark_conf:
        spark:
          jars:
            packages: io.delta:delta-core_2.11:0.6.1,com.microsoft.azure:azure-eventhubs-spark_2.11:2.3.15
  pyspark_step_launcher:
    config:
      databricks_host: https://uksouth.azuredatabricks.net
      databricks_token:
        env: DATABRICKS_TOKEN
      local_pipeline_package_path: .
      run_config:
        cluster:
          new:
            cluster_log_conf:
              dbfs:
                destination: dbfs:/cluster_logs
            nodes:
              instance_pool_id: some-instance-pool-id
            size:
              num_workers: 1
            spark_version: 6.6.x-scala2.11
        libraries:
        - pypi:
            package: dagster-azure==0.8.2
        - pypi:
            package: pyyaml==5.3.1
        - pypi:
            package: python-dotenv==0.13.0
        - maven:
            coordinates: com.microsoft.azure:azure-eventhubs-spark_2.11:2.3.15
      staging_prefix: /dagster-staging
      storage:
        adls2:
          secret_scope: foo
          storage_account_key_key: foo-key
          storage_account_name: foo
solids:
  create_table:
    config:
      table: access
storage:
  adls2:
    config:
      adls2_file_system: dagster
```

### CODE REVIEW

**Comment by natekupp** [comment]
Related: https://github.com/dagster-io/dagster/issues/2348

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py [hunk 1]
```diff
@@ -45,7 +50,7 @@
         'staging_prefix': Field(
             StringSource,
             is_required=False,
-            default_value='dagster_staging',
+            default_value='/dagster_staging',
```
Comment (sryza) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:53]: Not a huge deal, but because these changes are independent, it would be preferable to submit them as two separate PRs.

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_step_main.py [hunk 2]
```diff
@@ -102,6 +102,23 @@ def setup_storage(step_run_ref):
...
 
+def setup_environment(step_run_ref):
+    '''Setup any environment variables required by the run.
+
+    Extract any secrets in the run config and export them as environment variables.
+
+    This is important for any `StringSource` config since the environment variables
+    won't ordinarily be available in the Databricks execution environment.
+    '''
+    for secret in step_run_ref.run_config.get('secrets', []):
```
Comment (sryza) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_step_main.py:113]: This would need to be `step_run_ref.run_config['resources'][databricks_resource_name].get('secrets')`, right?  And in theory there could be multiple databricks resources?

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py [hunk 3]
```diff
@@ -31,6 +35,7 @@
         'databricks_token': Field(
             StringSource, is_required=True, description='Databricks access token',
         ),
+        'secrets': define_databricks_secrets_config(),
```
Comment (sryza) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:38]: I think it's worth being a little more explicit in the name about the weird thing that's happening here, e.g. `secrets_to_env_variables` or something.

**Comment by sryza** [comment]
Oof, this is a tricky one.  It feels a little scary / brittle the way that this uses run_config - we're relying on it both before and after the step where the environment variables get populated into the config.  What would you think about instead having the databricks_step_launcher serialize the set of secret names alongside the StepRunRef?  That would save us from needing to introspect the run_config on the remote side before that processing step.

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py (hunk 1 above)
Comment (sd2k) [reply python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:53]: I'll create a new PR for this change and the log message fix in the other commit :+1:

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py (hunk 3 above)
Comment (sd2k) [reply python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py]: Fixed.

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_step_main.py (hunk 2 above)
Comment (sd2k) [reply python_modules/libraries/dagster-databricks/dagster_databricks/databricks_step_main.py:113]: Good spot. I'd glossed over the fact that the name `pyspark_step_launcher` is chosen by the user rather than the library; that will need improving in the storage setup as well. Let's discuss in the main comment thread, since I agree we shouldn't be touching `run_config` like this here!

**Comment by sd2k** [comment]
Hmm yeah very good point - as I mentioned in the comment on the config name, I'd forgotten that there's no guarantee that the resource will even be named `pyspark_step_launcher`. I think it makes a lot of sense for the launcher to extract the config it needs and serialize it, perhaps even as a `DatabricksSetup` class which knows how to set itself up, so the `main` script has to do a bare minimum amount of work.

**Comment by sd2k** [comment]
I've implemented this using a `DatabricksConfig` class which is created and serialized by the step launcher, and contains the relevant storage + environment config. This is then loaded and set up by `databricks_step_main`, which no longer has to concern itself with the run_config.

It's somewhat easier to test with #2882 and #2883 merged in!

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py [hunk 4]
```diff
@@ -197,3 +216,69 @@ def _internal_dbfs_path(self, run_id, step_key, filename):
         """Scripts running on Databricks should access DBFS at /dbfs/."""
         path = "/".join([self.staging_prefix, run_id, step_key, os.path.basename(filename)])
         return "/dbfs/{}".format(path)
+
+
+class DatabricksConfig:
+    def __init__(self, storage, secrets):
```
Comment (sryza) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:222]: Would you mind including some docstring here explaining the purpose of this object and why we don't use the "regular" Dagster config system for it?

**Comment by sryza** [comment]
This looks good to me once we have some docstring for DatabricksConfig and get it passing the build.

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py (hunk 4 above)
Comment (sd2k) [reply python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:222]: I've added one now, hopefully it addresses everything.
